manual/zlib-flate.1.in
.TH ZLIB-FLATE "1" "" "zlib-flate from qpdf version @PROJECT_VERSION@" "User Commands"
.SH NAME
zlib-flate \- raw zlib compression program
.SH SYNOPSIS
.B zlib-flate
\fI-compress | -uncompress\fR
.SH DESCRIPTION
The zlib-flate program is part of the qpdf package.
.PP
The zlib-flate program reads from standard input and writes to
standard output either compressing or uncompressing its input using raw
zlib compression.  It can be used to uncompress or compress raw PDF
streams or other data that is compressed with raw zlib compression.
This program is provided primarily as a debugging tool, though it
could be used for other purposes, such as being called from a script
that creates simple PDF files.
.PP
This program should not be used as a general purpose compression
tool.  Use something like gzip(1) instead.
.PP
For details about qpdf, please see the qpdf manual, which can be found
at https://qpdf.readthedocs.io.
.SH "SEE ALSO"
qpdf(1), gzip(1)


ChangeLog
2022-03-08  Jay Berkenbilt  <ejb@ql.org>

	* 10.6.3: release

	* Use Windows 2022 github runners and therefore Visual Studio 2022
	to create Windows distributions

	* Fix DLL export issue with mingw (Windows)

2022-03-07  Jay Berkenbilt  <ejb@ql.org>

	* Minor internal changes to assist with building in other
	environments: rename internal bits.icc to qpdf/bits_functions.hh
	(not part of public API), enforce reordering of header files to
	prevent jpeglib.h from interfering with other headers, remove an
	unused header that was accidentally added in 10.6.0 but never
	referenced by any code.

	* Make build work and tests work when NDEBUG is defined. This
	involved a few changes to some test files but no changes to any
	library code.

2022-02-25  Jay Berkenbilt  <ejb@ql.org>

	* Bug fix in JSON parser: accept \/ in a string as valid input per
	JSON spec even though we don't translate / to \/ on output.

2022-02-22  Jay Berkenbilt  <ejb@ql.org>

	* Recognize PDF strings explicitly marked as UTF-8 as allowed by
	the PDF 2.0 spec. Fixes #654.

2022-02-18  Jay Berkenbilt  <ejb@ql.org>

	* Bug fix: when generating appearance streams, the font size was
	substituted incorrectly from /DA if Tf was absent or the number
	preceding Tf was out of range. Fixes #655.

2022-02-16  Jay Berkenbilt  <ejb@ql.org>

	* 10.6.2: release

2022-02-15  Jay Berkenbilt  <ejb@ql.org>

	* Fix asymmetrical logic between
	QPDFObjectHandle::newUnicodeString() and
	QPDFObjectHandle::getUTF8Val(). The asymmetrical logic didn't
	matter before fixing the PDF Doc transcoding bugs.

	* When analyzing PDF strings, recognize UTF-16LE as UTF-16. The
	PDF spec only allows UTF-16BE, but most readers seem to allow
	both. Fixes #649.

	* Bug fix: 10.6.0 inadvertently removed an unknown/undocumented
	CLI parsing feature, which has been restored in 10.6.2. Fixes #652.

	* Don't map 0x18 through 0x1f, 0x7f, 0x9f, or 0xad as fixed points
	when transcoding UTF-8 to PDFDoc. These code points have different
	meanings in those two encoding systems. Fixes #650.

2022-02-11  Jay Berkenbilt  <ejb@ql.org>

	* 10.6.1: release

	* Fix some compilation issues from use of abs without including
	proper headers.

2022-02-09  Jay Berkenbilt  <ejb@ql.org>

	* 10.6.0: release

	* Fix one more PDF doc encoding omission: 0xAD is also undefined.
	Fixes #637.

2022-02-08  Jay Berkenbilt  <ejb@ql.org>

	* Bug fix: when splitting pages with --split-pages or selecting
	pages with --pages, set the output PDF version to the maximum of
	all the input PDF versions. This is a fix to QPDFJob. If you are
	creating output PDF files yourself from multiple inputs, you will
	need to code the same thing. The new PDFVersion object, its
	updateIfGreater() method, and the new QPDF and QPDFWriter methods
	described below make this very easy to do. Fixes #610.

	* Add new class PDFVersion for more convenient comparison of PDF
	version numbers from the %!PDF header.

	* Add QPDF::getVersionAsPDFVersion() to return the PDF version and
	extension together as a PDFVersion object instead of a string.

	* Add a QPDFWriter::setMinimumPDFVersion() that takes a PDFVersion
	object.

2022-02-06  Jay Berkenbilt  <ejb@ql.org>

	* Pl_Buffer and QPDFWriter: add getBufferSharedPointer(), which
	turns a PointerHolder<Buffer> but will return a
	std::shared_ptr<Buffer> in qpdf 11.

	* From m-holger: add getKeyIfDict(), which calls getKey for
	dictionaries and returns null if called on null. This is for
	easier access to optional, lower-level dictionaries.

2022-02-05  Jay Berkenbilt  <ejb@ql.org>

	* Add several new accessors to QPDFObjectHandle: the bool
	getValueAsX(X&) accessors allow an alternative way to retrieve
	values from QPDFObjectHandle objects and can result in more
	concise code in many situations. Thanks to m-holger for the
	contribution.

	* Add qpdf_oh_new_binary_unicode_string and
	qpdf_oh_get_binary_utf8_value to the C API. This makes it possible
	to handle UTF-8-encoded strings with embedded NUL characters. Thanks
	to m-holger for the contribution.

	* Add a global user-defined string literal "_qpdf" as a shorthand
	for QPDFObjectHandle::parse, allowing you to create
	QPDFObjectHandle objects with

	QPDFObjectHandle oh = "<</Some (PDF)>>"_qpdf;

	* Expose QPDF::emptyPDF to the C API as qpdf_empty_pdf()

	* Add comments letting people know that the version string
	returned by QPDF::QPDFVersion and qpdf_get_qpdf_version is static.

	* Add QUtil::make_unique_cstr to return a std::unique_ptr<char[]>
	as an alternative to QUtil::copy_string and
	QUtil::make_shared_cstr.

2022-02-04  Jay Berkenbilt  <ejb@ql.org>

	* New preprocessor symbols QPDF_MAJOR_VERSION, QPDF_MINOR_VERSION,
	QPDF_PATCH_VERSION as numbers and QPDF_VERSION as a string. These
	can be used for feature testing in code. These are in qpdf/DLL.h,
	which is included by every header that adds to the public API.
	Since these constants are introduced in version 10.6, it's
	important for them to be in a header that everyone already
	includes so you don't have to try to include a header that won't
	be there.

	* PointerHolder: add a get() method and a use_count() method for
	compatibility with std::shared_ptr. In qpdf 11, qpdf's APIs will
	switch to using std::shared_ptr instead of PointerHolder, though
	there will be a PointerHolder class with a backward-compatible
	API. To ease the transition, we are adding get() now with the same
	semantics as std::shared_ptr's get. Note that there is a
	difference in behavior: const PointerHolder has always behaved
	incorrectly. const PointerHolder objects only returned const
	pointers. This is wrong. If you want a const pointer, use
	PointerHolder<T const>. A const PointerHolder just shouldn't allow
	its pointer to be reassigned. The new get() method behaves
	correctly in that calling get() on a const PointerHolder to a
	non-const pointer returns a non-const pointer. This is the way
	regular pointers behave.

2022-02-01  Jay Berkenbilt  <ejb@ql.org>


zlib-flate/zlib-flate.cc
#include <qpdf/Pl_Flate.hh>
#include <qpdf/Pl_StdioFile.hh>
#include <qpdf/QUtil.hh>
#include <qpdf/QPDF.hh>

#include <stdio.h>
#include <string.h>
#include <iostream>
#include <stdlib.h>
#include <fcntl.h>

static char const* whoami = 0;

void usage()
{
    std::cerr << "Usage: " << whoami << " { -uncompress | -compress[=n] }"
              << std::endl
              << "If n is specified with -compress, it is a"
              << " zlib compression level from" << std::endl
              << "1 to 9 where lower numbers are faster and"
              << " less compressed and higher" << std::endl
              << "numbers are slower and more compressed"
              << std::endl;
    exit(2);
}

int main(int argc, char* argv[])
{
    if ((whoami = strrchr(argv[0], '/')) == NULL)
    {
        whoami = argv[0];
    }
    else
    {
        ++whoami;
    }
    // For libtool's sake....
    if (strncmp(whoami, "lt-", 3) == 0)
    {
        whoami += 3;
    }

    if ((argc == 2) && (strcmp(argv[1], "--version") == 0))
    {
        std::cout << whoami << " from qpdf version "
                  << QPDF::QPDFVersion() << std::endl;
        exit(0);
    }

    if (argc != 2)
    {
        usage();
    }

    Pl_Flate::action_e action = Pl_Flate::a_inflate;

    if ((strcmp(argv[1], "-uncompress") == 0))
    {
        // okay
    }
    else if ((strcmp(argv[1], "-compress") == 0))
    {
        action = Pl_Flate::a_deflate;
    }
    else if ((strncmp(argv[1], "-compress=", 10) == 0))
    {
        action = Pl_Flate::a_deflate;
        int level = QUtil::string_to_int(argv[1] + 10);
        Pl_Flate::setCompressionLevel(level);
    }
    else
    {
        usage();
    }

    QUtil::binary_stdout();
    QUtil::binary_stdin();
    auto out = make_pointer_holder<Pl_StdioFile>("stdout", stdout);
    auto flate = make_pointer_holder<Pl_Flate>("flate", out.get(), action);
    bool warn = false;
    flate->setWarnCallback([&warn](char const* msg, int code) {
        warn = true;
        std::cerr << whoami << ": WARNING: zlib code " << code
                  << ", msg = " << msg << std::endl;
    });

    try
    {
        unsigned char buf[10000];
        bool done = false;
        while (! done)
        {
            size_t len = fread(buf, 1, sizeof(buf), stdin);
            if (len <= 0)
            {
                done = true;
            }
            else
            {
                flate->write(buf, len);
            }
        }
        flate->finish();
    }
    catch (std::exception& e)
    {
        std::cerr << whoami << ": " << e.what() << std::endl;
        exit(2);
    }

    if (warn)
    {
        exit(3);
    }
    return 0;
}
